ociclient/client.go

```diff
@@ -552,3 +565,215 @@ func AddKnownMediaTypesToCtx(ctx context.Context, mediaTypes []string) context.C
+	}
+
+	v2ManifestDesc, v2ManifestBytes, err := createV2Manifest(configDesc, layers)
+	if err != nil {
+		return ocispecv1.Descriptor{}, fmt.Errorf("unable to create v2 manifest: %w", err)
+	}
+
+	err = c.cache.Add(configDesc, io.NopCloser(bytes.NewReader(configBytes)))
+	if err != nil {
+		return ocispecv1.Descriptor{}, fmt.Errorf("unable to write config blob to cache: %w", err)
+	}
+
+	err = c.cache.Add(v2ManifestDesc, io.NopCloser(bytes.NewReader(v2ManifestBytes)))
+	if err != nil {
+		return ocispecv1.Descriptor{}, fmt.Errorf("unable to write manifest blob to cache: %w", err)
+	}
+
+	return v2ManifestDesc, nil
+}
+
+func createV2Manifest(configDesc ocispecv1.Descriptor, layers []ocispecv1.Descriptor) (ocispecv1.Descriptor, []byte, error) {
+	v2Manifest := ocispecv1.Manifest{
+		Versioned: specs.Versioned{
+			SchemaVersion: 2,
+		},
+		Config: configDesc,
+		Layers: layers,
+	}
+
+	marshaledV2Manifest, err := json.MarshalIndent(v2Manifest, "", "   ")
+	if err != nil {
+		return ocispecv1.Descriptor{}, nil, fmt.Errorf("unable to marshal manifest: %w", err)
+	}
+
+	v2ManifestDesc := ocispecv1.Descriptor{
+		MediaType: ocispecv1.MediaTypeImageManifest,
+		Digest:    digest.FromBytes(marshaledV2Manifest),
+		Size:      int64(len(marshaledV2Manifest)),
+	}
+
+	return v2ManifestDesc, marshaledV2Manifest, nil
+}
+
+func createConfig(v1Manifest *v1Manifest, diffIDs []digest.Digest, history []ocispecv1.History) (ocispecv1.Descriptor, []byte, error) {
+	var config map[string]*json.RawMessage
+	if err := json.Unmarshal([]byte(v1Manifest.History[0].V1Compatibility), &config); err != nil {
+		return ocispecv1.Descriptor{}, nil, fmt.Errorf("unable to unmarshal config from v1 history: %w", err)
+	}
+
+	delete(config, "id")
+	delete(config, "parent")
+	delete(config, "Size") // Size is calculated from data on disk and is inconsistent
+	delete(config, "parent_id")
+	delete(config, "layer_id")
+	delete(config, "throwaway")
+
+	rootfs := ocispecv1.RootFS{
+		Type:    "layers",
+		DiffIDs: diffIDs,
+	}
+
+	config["rootfs"] = utils.RawJSON(rootfs)
+	config["history"] = utils.RawJSON(history)
+
+	marshaledConfig, err := json.Marshal(config)
+	if err != nil {
+		return ocispecv1.Descriptor{}, nil, fmt.Errorf("unable to marshal config: %w", err)
+	}
+
+	configDesc := ocispecv1.Descriptor{
+		MediaType: ocispecv1.MediaTypeImageConfig,
+		Digest:    digest.FromBytes(marshaledConfig),
+		Size:      int64(len(marshaledConfig)),
+	}
+
+	return configDesc, marshaledConfig, nil
+}
+
+// isEmptyLayer returns whether the v1 compatibility history describes an
+// empty layer. A return value of true indicates the layer is empty,
+// however false does not indicate non-empty.
+func isEmptyLayer(h *v1History) bool {
+	if h.ThrowAway != nil {
+		return *h.ThrowAway
+	}
+	if h.Size != nil {
+		return *h.Size == 0
+	}
+
+	// If no `Size` or `throwaway` field is given, then it cannot be determined whether the layer is empty
+	// from the history, return false
+	return false
```

Comment:
shouldn't we throw an error or is this a valid state

Reply:
I've not found any formal spec that describe whether these fields must exist or not. But from what I've seen in the Docker implementation, `throwAway` is optional (https://github.com/moby/moby/blob/master/distribution/pull_v2.go#L524). For the `size` attribute I can only assume the same. 

I think the whole logic should be interpreted as: "If clients which pushed the content already made the effort to indicate that something is empty, we can safely throw it away. For all other cases we copy every layer."